simulator/syncer/resource.go

```diff
@@ -0,0 +1,122 @@
+}
+
+// Event is a type of events that occur in the source cluster.
+type Event int
+
+const (
+	Add Event = iota
+	Update
+)
+
+// MandatoryMutatingFunctions is a map of GroupVersionResource to MutatingFunction.
+// Outside users must call them in their own mutating functions.
+// Note: MandantoryMutatingFunctions overwrites pods' ServiceAccountName with default.
+var MandatoryMutatingFunctions = map[schema.GroupVersionResource]MutatingFunction{
+	{Group: "", Version: "v1", Resource: "persistentvolumes"}: mutatePV,
+	{Group: "", Version: "v1", Resource: "pods"}:              mutatePods,
+}
+
+// MandatoryFilteringFunctions is a map of GroupVersionResource to FilteringFunction.
+// Outside users must call them in their own filtering functions.
+var MandatoryFilteringFunctions = map[schema.GroupVersionResource]FilteringFunction{
+	{Group: "", Version: "v1", Resource: "pods"}: filterPods,
+}
+
+// FilteringFunction is a function that filters a resource.
+// If it returns false, the resource will not be imported.
+type FilteringFunction func(ctx context.Context, resource *unstructured.Unstructured, clients *Clients, event Event) (bool, error)
+
+// MutatingFunction is a function that mutates a resource before importing it.
+type MutatingFunction func(ctx context.Context, resource *unstructured.Unstructured, clients *Clients, event Event) (*unstructured.Unstructured, error)
+
+func mutatePV(ctx context.Context, resource *unstructured.Unstructured, clients *Clients, _ Event) (*unstructured.Unstructured, error) {
+	var pv v1.PersistentVolume
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(resource.UnstructuredContent(), &pv)
+	if err != nil {
+		return nil, err
+	}
+
+	if pv.Status.Phase == v1.VolumeBound {
+		// PersistentVolumeClaims's UID is changed in a destination cluster when importing from a source cluster,
+		// and thus we need to update the PVC UID in the PersistentVolume.
+		// Get PVC of pv.Spec.ClaimRef.Name.
+		pvc, err := clients.SrcDynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "",
+			Version:  "v1",
+			Resource: "persistentvolumeclaims",
+		}).Namespace(pv.Spec.ClaimRef.Namespace).Get(ctx, pv.Spec.ClaimRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		pv.Spec.ClaimRef.UID = pvc.GetUID()
+	}
+
+	modifiedUnstructed, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&pv)
+	return &unstructured.Unstructured{Object: modifiedUnstructed}, err
+}
+
+func mutatePods(_ context.Context, resource *unstructured.Unstructured, _ *Clients, _ Event) (*unstructured.Unstructured, error) {
+	var pod v1.Pod
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(resource.UnstructuredContent(), &pod)
+	if err != nil {
+		return nil, err
+	}
+
+	// Pods must have the default ServiceAccount because ServiceAccount is not synced.
+	// But static pods must not have a ServiceAccount.
+	if pod.Spec.ServiceAccountName != "" {
+		pod.Spec.ServiceAccountName = "default"
+	}
```

Comment:
```suggestion
	
	// We always make it empty so that the default service account is used.
	// (We don't sync a service account because it's not related to the scheduling)
	pod.Spec.ServiceAccountName = ""
```

Reply:
If not static pods are applied without `ServiceAccountName`, the resource creation fails.

Reply:
Does it?

```bash
sanposhiho@Kenseis-MacBook-Pro workspace % cat pod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
sanposhiho@Kenseis-MacBook-Pro workspace % kubectl apply -f pod.yaml
pod/nginx created
```

Reply:
Just applying it via kubectl is okay. But applying not static pods without ServiceAccoutName in the syncer fails.

I don't know why it causes, but you can repro by the following steps:

- Do the fix you suggest.
- Start the simulator with the syncer.
- Create `foo` ServiceAccount and a pod that has `foo` ServiceAccount in the source simulator.

```
kwokctl kubectl apply -f - <<EOF
apiVersion: v1
kind: ServiceAccount
metadata:
  name: foo
---
apiVersion: v1
kind: Pod
metadata:
  name: nginx-1
spec:
  serviceAccountName: foo
  containers:
  - name: nginx
    image: nginx:1.14.2
EOF
```
- In the simulator cluster, you won't see the pod. The simulator server's log says:

```
E0828 09:24:53.405157       1 syncer.go:230] "Failed to create resource on destination cluster" err="failed to create resource: pods \"nginx-1\" is forbidden: error looking up service account default/foo: serviceaccount \"foo\" not found"
```

Reply:
What if we remove `ManagedFields` too?

Reply:
> What if we remove `ManagedFields` too?

The result is the same.

Reply:
This works:

```
	pod.Spec.ServiceAccountName = ""
	pod.Spec.DeprecatedServiceAccount = ""
```